test/common/upstream/deferred_cluster_initialization_test.cc

```diff
@@ -516,6 +516,108 @@ TEST_P(EdsTest, ShouldMergeAddingHosts) {
                                  {1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000}));
 }
 
+TEST_P(EdsTest, ShouldNotMergeAddingHostsForDifferentClustersWithSameName) {
+  const std::string bootstrap_yaml = R"EOF(
+    static_resources:
+      clusters:
+      - name: eds
+        connect_timeout: 0.250s
+        lb_policy: ROUND_ROBIN
+        load_assignment:
+          cluster_name: bootstrap_cluster
+          endpoints:
+          - lb_endpoints:
+            - endpoint:
+                address:
+                  socket_address:
+                    address: 127.0.0.1
+                    port_value: 60000
+    )EOF";
+
+  auto bootstrap = parseBootstrapFromV3YamlEnableDeferredCluster(bootstrap_yaml);
+  create(bootstrap);
+
+  const std::string eds_cluster_yaml = R"EOF(
+      name: cluster_1
+      connect_timeout: 0.25s
+      lb_policy: RING_HASH
+      eds_cluster_config:
+        service_name: fare
+        eds_config:
+          api_config_source:
+            api_type: REST
+            transport_api_version: V3
+            cluster_names:
+            - eds
+            refresh_delay: 1s
+    )EOF";
+
+  EXPECT_CALL(factory_, create(_))
+      .Times(2)
+      .WillRepeatedly(
+          testing::Invoke([this](Config::ConfigSubscriptionFactory::SubscriptionData& data) {
+            callbacks_ = &data.callbacks_;
+            return std::make_unique<NiceMock<Envoy::Config::MockSubscription>>();
+          }));
+
+  EXPECT_EQ(readGauge("thread_local_cluster_manager.test_thread.clusters_inflated"), 0);
+  EXPECT_TRUE(cluster_manager_->addOrUpdateCluster(
+      parseClusterFromV3Yaml(eds_cluster_yaml, getEdsClusterType()), "version1"));
+
+  envoy::config::endpoint::v3::ClusterLoadAssignment cluster_load_assignment;
+  cluster_load_assignment.set_cluster_name("fare");
+  for (int i = 1; i < 11; ++i) {
+    addEndpoint(cluster_load_assignment, 1000 * i);
+    doOnConfigUpdateVerifyNoThrow(cluster_load_assignment);
+  }
+
+  auto initiailization_instance =
+      cluster_manager_->clusterInitializationMap().find("cluster_1")->second;
+  EXPECT_NE(nullptr, initiailization_instance->load_balancer_factory_);
+  // RING_HASH lb policy requires Envoy re-create the load balancer when the cluster is updated.
+  EXPECT_TRUE(initiailization_instance->load_balancer_factory_->recreateOnHostChange());
+
+  const std::string new_eds_cluster_yaml = R"EOF(
+      name: cluster_1
+      connect_timeout: 0.25s
+      lb_policy: ROUND_ROBIN
+      eds_cluster_config:
+        service_name: new_fare
```

Comment:
Is it really needed to change the EDS service-name?

Reply:
From test's view, this change (and the LB change) is/are want to tell this cluster is complete different with previous same name cluster.
So, I think this is necessary.

Reply:
I think that making a small modification (e.g., just the LB) makes more sense.
When I tried to understand where the issue is coming from, I couldn't understand if this is a required change or not. Keeping the test minimal just helps making it very explicit. WDYT?

Reply:
Basically, I has no reason to refute it. I am convinced by you now. Will update it.